README
bestfit - Overview
==================

Bestfit is a small program to determine which files that should be put
on a CD (or other media), so that as little space as possible is wasted.
It is very easy to use: you specify files on the command line, and bestfit
prints the names of those that were selected. Alternatively, bestfit
can execute a command for each selected file (e.g. to move them to a
different directory).

This problem is also known as the 0-1 knapsack problem. The algorithm
implemented by bestfit solves instances of it optimally in theta(n*W)
time, where n is number of files to choose between and W the amount of
free space. This may sound bad but since W is number of blocks and not
bytes, this algorithm is not that inefficient. Try it yourself and see.

Due to the nature of this algorithm, bestfit uses a lot of memory -
approximately 1.5M per file specified on command line. As long as
you have enough swap space, this is usually not a problem since
bestfit doesn't use all at the same time.

Bestfit does (at this time) not use any external program such as mkisofs
or mkhybrid to calculate the size of files. Instead the file and directory
sizes and rounded up towards the nearest 2048-byte multiple and added
together. (2048 bytes is the block size on ISO images.)

Bestfit can not account for all oddities with ISO images (for instance,
there are multiple path tables, padding, etc which is not easy to
calculate), therefore exact results can not always be produced. Bestfit
may waste more space than necessary, or even tell you that files that
won't fit actually fits. (I consider the latter to be a bug though.
If you encounter it, please mail me.) In other words, manually selecting
files and checking that they fit using mkisofs may in some cases give
you better results than using bestfit.

License
=======

Bestfit is copyright (C) 2001  Oskar Liljeblad

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Requirements
============

* GNU libc 2.0 or later
  Bestfit uses some GNU extensions of libc (such as getopt_long).

Installation
============

Compile bestfit by running `configure' and then `make'.
This should produce an executable `bestfit' in the src
directory, which can be used right away.

If you would like to install bestfit on your system, run
`make install'. This will copy the executable and the man page
`bestfit.1' into the appropriate directories (usually
/usr/local/bin and /usr/local/share/man/man1, respecively).

For further details regarding configure and the make directives,
see the INSTALL file.

Usage and Examples
==================

Run `bestfit --help' to display a list of options, or
see the man page (`man ./docs/bestfit.1'). The man page
also contains some examples on how to use bestfit.

Known bugs
==========

None at the moment.

Possible things to do
=====================

See the TODO file.

Author and Feedback
===================

The author of bestfit and this document, Oskar Liljeblad,
can be reached via email address

  oskar@osk.mine.nu
  
Please send bug reports, suggestions, ideas or comments
in general to me.

Homepage
========

The latest version of bestfit can be downloaded from

  http://www.student.lu.se/~nbi98oli/

-


NEWS
2001-10-01

The total block variable is now 64 bits, fixing possible
overflow problems.
The author has changed email-address to oskar@osk.mine.nu.
Added patch from Steven Enderle to add support for FreeBSD.
Updated to automake 1.4p4.

2001-04-23

First public release (0.1.0).

-


src/bestfit.h
/* bestfit.h - Global stuff for bestfit
 * 
 * Copyright (C) 1998-2001  Oskar Liljeblad
 *
 * This file is part of bestfit.
 *
 * This software is copyrighted work licensed under the terms of the
 * GNU General Public License. Please consult the file "LICENSE" for
 * details.
 */

#ifndef __BESTFIT_H__
#define __BESTFIT_H__

#include "../config.h"
#include "support.h"

#ifdef HAVE_STDINT_H
#include <stdint.h>
#endif

/* Number of 2048 byte block on a 74-min CD */
#define ISO_BLOCKS_74_MIN_CD	333000
/* Number of 2048 byte block on a 80-min CD */
#define ISO_BLOCKS_80_MIN_CD	360000
/* Number of 2048 byte block on a 4,7 GB Single Layer DVD */
#define ISO_BLOCKS_DVD_SL	2295104
/* Number of 2048 byte block on a 8,5 GB Double Layer DVD */
#define ISO_BLOCKS_DVD_DL	4171712
/* Default number of blocks to reserve on CDs */
#define DEFAULT_RESERVED_BLOCKS 200
/* The default (and unmodifyable) block size */
#define DEFAULT_BLOCK_SIZE  	2048

typedef void (*KnapsackStatusCallback)(uint32_t current, uint32_t total);

/* sizescan.c */
uint64_t get_simple_file_size(char *name, int blocksize);
char *get_error_file();

/* knapsack.c */
bool *knapsack(uint32_t *weights, int items, int space,
    	KnapsackStatusCallback status_cb, int reportdelay);

#endif /* __BESTFIT_H__ */


src/support.c
/* support.c - Support routines and definitions
 * 
 * Copyright (C) 1998-2001  Oskar Liljeblad
 *
 * This file is part of bestfit.
 *
 * This software is copyrighted work licensed under the terms of the
 * GNU General Public License. Please consult the file "LICENSE" for
 * details.
 */

#include "support.h"

#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef HAVE_STDINT_H
#include <stdint.h>
#endif
#ifdef STDC_HEADERS
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#endif

/* xstrtoi:
 * More simple version of strtol.
 */
bool
xstrtoi(char *str, int *result)
{
    char *endptr;
    long int lresult;

    lresult = strtol(str, &endptr, 10);
    if (str[0] == '\0' || endptr[0] != '\0')
    	return FALSE;

    *result = (int) lresult;
    return TRUE;
}

/* xmalloc:
 * Safe malloc, that will never return NULL.
 */
void *
xmalloc(size_t size)
{
    void *mem = malloc(size);
    assert_memory(mem != NULL);
    return mem;
}

/* Make sure that a memory allocation operation has succeeded.
 * Otherwise we print an message, and exit with non-zero status.
 */
void
assert_memory(bool expression)
{
    if (!expression) {
    	fprintf(stderr, "%s\n", strerror(ENOMEM));
	exit(1);
    }
}

/* is_directory:
 * Return TRUE if file exists, is readable and is a directory.
 */
bool
is_directory(char *filename)
{
    struct stat statbuf;
    if (stat(filename, &statbuf) == -1)
	return FALSE;
    return S_ISDIR(statbuf.st_mode);
}


src/support.h
/* support.h - Support routines and definitions
 * 
 * Copyright (C) 1998-2001  Oskar Liljeblad
 *
 * This file is part of bestfit.
 *
 * This software is copyrighted work licensed under the terms of the
 * GNU General Public License. Please consult the file "LICENSE" for
 * details.
 */

#ifndef __SUPPORT_H__
#define __SUPPORT_H__

#include "../config.h"

#ifdef HAVE_STDINT_H
#include <stdint.h>
#endif
#ifdef STDC_HEADERS
#include <stdlib.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif

/* bool:
 * Boolean type (use TRUE and FALSE).
 */
typedef int8_t bool;

/* FALSE:
 * TRUE:
 * Boolean values.
 */
#define FALSE	    0
#define TRUE	    (!FALSE)

bool xstrtoi(char *str, int *result);
void *xmalloc(size_t size);
void assert_memory(bool expression);
bool is_directory(char *filename);

#endif /* __SUPPORT_H__ */


src/knapsack.c
/* knapsack.c - Algorithm for solving the 0-1 knapsack problem
 * 
 * Copyright (C) 1998-2001  Oskar Liljeblad
 *
 * This file is part of bestfit.
 *
 * This software is copyrighted work licensed under the terms of the
 * GNU General Public License. Please consult the file "LICENSE" for
 * details.
 */

#include "bestfit.h"

#define C(i,j)  (c[(i)*(W+1) + (j)])

#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#ifdef HAVE_LIMITS_H
#include <limits.h>
#endif

/* Run the knapsack algorithm.
 *
 * /weights/ is an array with at least /items/ item weights.
 * /space/ is the maximum total size.
 * /status_cb/ is a function that will be called every /reportdelay/ units.
 * (A /reportdelay/ of MAX_INT and /status_cb/ of NULL disables reporting.)
 *
 * This function allocates and returns an array where cell /i/ indicates
 * whether item /i/ is included in the knapsack (TRUE) or not (FALSE).
 * This array should be freed with /free/ when no longer needed.
 */
bool *
knapsack(uint32_t *w, int n, int W, KnapsackStatusCallback status_cb,
    	int reportdelay)
{
    register int i;
    register int j;
    int *c;
    int minsize;
    bool *results;

    c = xmalloc(sizeof(int) * (n+1) * (W+1));
    results = xmalloc(sizeof(bool) * n);

    minsize = INT_MAX;

    for (j = 0; j <= W; j++) {
    	C(0, j) = 0;
    }
    for (i = 1; i <= n; i++) {
    	minsize = MIN(minsize, w[i-1]);
	minsize = MIN(minsize, W);
	minsize = MAX(minsize, 1);

    	for (j = 0; j < minsize; j++) {
    	    C(i, j) = 0;
	}

	for (j = minsize; j <= W; j++) {
	    if (j % reportdelay == 0 && status_cb != NULL)
	    	status_cb(j+W*(i-1), W*n);

	    if (w[i-1] <= j && w[i-1] + C(i-1, j-w[i-1]) > C(i-1, j)) {
	    	C(i, j) = w[i-1] + C(i-1, j-w[i-1]);
	    } else {
    	    	C(i, j) = C(i-1, j);
	    }
	}
    }

    j = W;
    for (i = n; i > 0; i--) {
	if (w[i-1] <= j && C(i, j) == w[i-1] + C(i-1, j-w[i-1])) {
	    results[i-1] = TRUE;
	    j -= w[i-1];
	} else {
	    results[i-1] = FALSE;
	}
    }

    return results;
}


src/sizescan.c
/* sizescan.c - Routines for figuring out size of files
 * 
 * Copyright (C) 1998-2001  Oskar Liljeblad
 *
 * This file is part of bestfit.
 *
 * This software is copyrighted work licensed under the terms of the
 * GNU General Public License. Please consult the file "LICENSE" for
 * details.
 */

#include "bestfit.h"

#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_DIRENT_H
#include <dirent.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef HAVE_STDINT_H
#include <stdint.h>
#endif
#ifdef HAVE_LIMITS_H
#include <limits.h>
#endif
#ifdef STDC_HEADERS
#include <string.h>
#endif

static char errorfile[PATH_MAX+1];

/* If get_simple_file_size returned -1, this function will
 * return a pointer to a static buffer containing the name
 * of the file that caused the error.
 */
char *
get_error_file()
{
    return errorfile;
}

/* Return size of file denoted by /name/. If it is a directory,
 * scan it recursively and return its total size. All returned
 * sizes are rounded upwards to the nearest /blocksize/ multiple.
 *
 * -1 is returned if there was an error.
 */
uint64_t
get_simple_file_size(char *name, int blocksize)
{
    char namebuf[PATH_MAX+1];
    uint64_t size = 0;
    DIR *dir;
    struct dirent *ent;
    struct stat statbuf;

    if (stat(name, &statbuf) == -1) {
    	if (lstat(name, &statbuf) == -1) {
	    strcpy(errorfile, name);
    	    return -1;
	}
    }

    size = statbuf.st_size;
    if (size % blocksize != 0)
    	size = (size/blocksize + 1) * blocksize;

    if (S_ISDIR(statbuf.st_mode)) {
	if ((dir = opendir(name)) == NULL) {
	    strcpy(errorfile, name);
    	    return -1;
	}

   	errno = 0;
    	while ((ent = readdir(dir)) != NULL) {
   	    if (strcmp(ent->d_name, ".") != 0
	    	    && strcmp(ent->d_name, "..") != 0) {
    	    	uint64_t tsize;
	    	int needsslash = (name[strlen(name)-1] != '/');

    	    	/* If the new filename is too long, bail out.
		 * Try to make a reasonable name for the errorfile though.
		 */
    	    	if (strlen(name)+needsslash+strlen(ent->d_name) > PATH_MAX) {
	    	    strcpy(errorfile, name);
		    if (needsslash && strlen(name)+needsslash <= PATH_MAX)
		    	strcat(errorfile, "/");
		    if (strlen(name)+needsslash+2 <= PATH_MAX)
		    	strcat(errorfile, "..");
		    errno = ENAMETOOLONG;
		    return -1;
		}

    	    	strcpy(namebuf, name);
		if (needsslash)
	    	    strcat(namebuf, "/");
	    	strcat(namebuf, ent->d_name);

    	    	tsize = get_simple_file_size(namebuf, blocksize);
	    	if (tsize == -1) {
		    closedir(dir);
		    return -1;
		}
	    	size += tsize;
	    }
    	    errno = 0;
	}

	if (errno != 0)
    	    size = -1;
	closedir(dir);
    }

    return size;
}


src/bestfit.c
/* bestfit.c - The main source file
 * 
 * Copyright (C) 1998-2001  Oskar Liljeblad
 *
 * This file is part of bestfit.
 *
 * This software is copyrighted work licensed under the terms of the
 * GNU General Public License. Please consult the file "LICENSE" for
 * details.
 */

#include "bestfit.h"

#define _GNU_SOURCE

#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_LIMITS_H
#include <limits.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef STDC_HEADERS
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#endif

#include "getopt.h"

/* Verbosity level: 0 = quiet, 1 = normal, 2 = verbose */
static int verbosity = 1;
/* Program name: same as argv[0] */
static char *program_name;

/* VERSION_MESSAGE:
 * Message to display when the --version option is used.
 */
#define VERSION_MESSAGE \
PACKAGE " " VERSION "\n\
Written by Oskar Liljeblad.\n\
\n\
Copyright (C) 2001 Oskar Liljeblad\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"

/* HELP_MESSAGE:
 * Message to display when the --help option is used.
 */
#define HELP_MESSAGE "\
Usage: %s [OPTION]... FILE... [-e COMMAND...]\n\
Determine which FILEs should be put on a CD (or other disk) so that as\n\
little space as possible is wasted. Output the name of those files,\n\
or if --execute specified, execute a command for each.\n\
\n\
  -8, --80min               space is that of an empty 80 min CD (default)\n\
  -7, --74min               space is that of an empty 74 min CD\n\
  -5, --sldvd               space is that of an empty 4,7 GB DVD \n\
  -9, --dldvd               space is that of an empty 8,5 GB DVD \n\
  -d, --dirs-only           ignore non-directory arguments\n\
  -s, --space=BLOCKS        maximum amount of space allowed to use\n\
  -r, --reserve=BLOCKS      blocks to reserve (%d by default)\n\
  -e, --execute COMMAND...  execute commands for each selected file\n\
  -i, --invert              print/execute for non-fitting files instead\n\
  -v, --verbose             explain what is being done\n\
  -q, --quiet               output no diagnostic/information messages\n\
      --help                display this help and exit\n\
      --version             output version information and exit\n\
\n\
The block-size is fixed to 2048 bytes.\n\
\n\
All following arguments after --execute are taken to be arguments\n\
to the command to execute for each selected file. The string `{}'\n\
is replaced by the current file name being processed.\n\
\n\
Report bugs to <" BUG_EMAIL ">.\n"

static void print_knapsack_status(uint32_t current, uint32_t total);
static void verbose_printf(int minlvl, char *format, ...);
static void execute_command(char *file, char **argv);

static struct option long_options[] = {
    { "80min",	    	no_argument,	    NULL, '8' },
    { "74min",	    	no_argument,	    NULL, '7' },
    { "sldvd",	    	no_argument,	    NULL, '5' },
    { "dldvd",	    	no_argument,	    NULL, '9' },
    { "dirs-only",   	no_argument,	    NULL, 'd' },
    { "space",          required_argument,  NULL, 's' },
    { "reserve",    	required_argument,  NULL, 'r' },
    { "help",           no_argument,        NULL, 151 },
    { "version",        no_argument,        NULL, 150 },
    { "quiet",          no_argument,        NULL, 'q' },
    { "silent",         no_argument,        NULL, 'q' },
    { "invert",         no_argument,        NULL, 'i' },
    { "execute",        no_argument,        NULL, 'e' },
    { 0, 0, 0, 0 }
};

/* Print an error message to the standard error stream.
 */
static void
verbose_printf(int minlvl, char *format, ...)
{
    va_list args;

    va_start(args, format);
    if (verbosity >= minlvl) {
	vfprintf(stderr, format, args);
    	fflush(stderr);
    }
    va_end(args);
}

/* Called by knapsack function to display status information.
 */
static void
print_knapsack_status(uint32_t current, uint32_t total)
{
    verbose_printf(1, "  %4.1f%%\r", current*100.0/total);
}

/* Execute a command with specified arguments for the specified file.
 * Replace all occurences of {} in /argv/ with /file/.
 */
static void
execute_command(char *file, char **argv)
{
    pid_t child;

    child = fork();
    if (child == 0) {	/* in child */
    	int i, j, k;
	char *t;

    	/* Replace all occurences of {} with /file/ */
    	for (i = 1; argv[i] != NULL; i++) {
	    t = argv[i];
	    for (j = 0; (t = strstr(t, "{}")) != NULL; j++)
	    	t++;

    	    if (j > 0) {
    	    	t = xmalloc(strlen(argv[i]) - j*2 + j*strlen(file) + 1);
		k = 0;
	    	for (j = 0; argv[i][j] != 0; j++) {
		    if (argv[i][j] == '{' && argv[i][j+1] == '}') {
		    	strcpy(&t[k], file);
			k += strlen(file);
		    	j++;
		    } else {
		    	t[k++] = argv[i][j];
		    }
		}
		t[k] = '\0';
	    	argv[i] = t;
	    }
	}
    
    	execvp(argv[0], argv);
    	perror(program_name);
    	exit(1);
    } else if (child != -1) { /* in parent */
    	int status = 0;
    	if (waitpid(child, &status, 0) == -1) {
	    perror(program_name);
	    exit(1);
	}
	if (!WIFEXITED(status)) {
	    fprintf(stderr, "%s: child did not exit normally\n", program_name);
	    exit(1);
	}
    } else {
    	perror(program_name);
	exit(1);
    }
}

int
main (int argc, char **argv)
{
    int i;
    int j;
    int nonoptscount;
    uint64_t total;
    uint32_t used;
    uint32_t *weights;
    bool *results;
    char **nonopts;
    int space = ISO_BLOCKS_80_MIN_CD;
    int reserve = DEFAULT_RESERVED_BLOCKS;
    bool directories_only = FALSE;
    bool invert = FALSE;
    bool execute = FALSE;

    program_name = argv[0];
    nonopts = (argc <= 1 ? NULL : xmalloc(sizeof(char *) * (argc-1)));
    nonoptscount = 0;

    while (TRUE) {
    	int option_index = 0;
	i = getopt_long(argc, argv, "-8759dqver:s:", long_options, &option_index);
    	if (i == EOF)
	    break;

    	if (i == 'e') {
	    if (argc <= optind) {
	    	printf("%s: missing arguments for --execute\n", program_name);
		exit(1);
	    }
	    execute = TRUE;
	    break;
	}

    	switch (i) {
	case '8': /* --80min */
	    space = ISO_BLOCKS_80_MIN_CD;
	    break;

	case '7': /* --74min */
	    space = ISO_BLOCKS_74_MIN_CD;
	    break;
	
	case '5': /* --4,7 GB DVD */
	    space = ISO_BLOCKS_DVD_SL;
	    break;

	case '9': /* --8,5 GB DVD */
	    space = ISO_BLOCKS_DVD_DL;
	    break;

    	case 'd': /* --dirs-only */
	    directories_only = TRUE;
	    break;

    	case 'q': /* --quiet */
	    verbosity = 0;
	    break;

	case 'v': /* --verbose */
	    verbosity = 2;
	    break;

    	case 's': /* --space */
	    if (!xstrtoi(optarg, &space)) {
		fprintf(stderr, "%s: invalid integer: %s\n", program_name, optarg);
		exit(1);
	    }
    	    break;

    	case 'r': /* --reserve */
	    if (!xstrtoi(optarg, &reserve)) {
		fprintf(stderr, "%s: invalid integer: %s\n", program_name, optarg);
		exit(1);
	    }
    	    break;

    	case 'i': /* --invert */
	    invert = TRUE;
	    break;

	case 150: /* --version */
	    printf(VERSION_MESSAGE);
	    exit(0);

	case 151: /* --help */
	    printf(HELP_MESSAGE, program_name, DEFAULT_RESERVED_BLOCKS);
	    exit(0);

    	case 1: /* non-option arguments */
	    nonopts[nonoptscount++] = optarg;
	    break;

	case '?':
	default:
	    exit(1);
    	}
    }

    if (verbosity == 1 && !isatty(STDERR_FILENO))
    	verbosity = 0;

    if (nonoptscount == 0) {
	fprintf(stderr, "%s: missing arguments\n", program_name);
	fprintf(stderr, "Try `%s --help' for more information.\n", program_name);
	exit(1);
    }

    weights = xmalloc(sizeof(uint32_t) * nonoptscount);
    for (i = 0; i < nonoptscount; i++) {
    	uint64_t size;

	if (directories_only && !is_directory(nonopts[i])) {
	    verbose_printf(2, "ignored file `%s'\n", nonopts[i]);
    	    nonoptscount--;
    	    for (j = i; j < nonoptscount; j++)
	    	nonopts[j] = nonopts[j+1];
	    i--;
	    continue;
	}

        verbose_printf(2, "scanning `%s': ", nonopts[i]);
    	size = get_simple_file_size(nonopts[i], DEFAULT_BLOCK_SIZE);
	if (size == -1) {
	    verbose_printf(2, "failed\n");
	    fprintf(stderr, "%s: %s\n", get_error_file(), strerror(errno));
	    exit(1);
	}
    	verbose_printf(2, "%lld %s\n", size, (size == 1 ? "byte" : "bytes"));

    	if ((uint64_t) (size/DEFAULT_BLOCK_SIZE) > (uint64_t) UINT_MAX) {
	    verbose_printf(2, "`%s' has more than 2^32 block - ignoring\n", nonopts[i]);
    	    nonoptscount--;
    	    for (j = i; j < nonoptscount; j++)
	    	nonopts[j] = nonopts[j+1];
	    i--;
	    continue;
    	}

    	weights[i] = size/DEFAULT_BLOCK_SIZE;
    }

    if (nonoptscount == 0) {
	fprintf(stderr, "%s: no remaining arguments\n", program_name);
	exit(1);
    }

    space -= reserve;
    results = knapsack(weights, nonoptscount, space, print_knapsack_status,
    	    	    (verbosity == 0 ? INT_MAX : 10000));
    verbose_printf(1, "       \r");

    used = 0;
    total = 0;
    for (i = 0; i < nonoptscount; i++) {
    	if (invert ^ results[i]) {
	    if (execute) {
    	    	verbose_printf(2, "%s\n", nonopts[i]);
		execute_command(nonopts[i], &argv[optind]);
	    } else {
	    	printf("%s\n", nonopts[i]);
	    }
	    used += weights[i];
	} else if (invert) {
	    verbose_printf(2, "include `%s'\n", nonopts[i]);
	} else {
	    verbose_printf(2, "did not include `%s'\n", nonopts[i]);
	}
	total += weights[i];
    }

    verbose_printf(1, "used %d of %lld %s (%.1fM of %.1fM)\n",
    	    used,
    	    total,
	    (total == 1 ? "block" : "blocks"),
	    (used*DEFAULT_BLOCK_SIZE) / 1048576.0,
	    (total*DEFAULT_BLOCK_SIZE) / 1048576.0);

    verbose_printf(1, "wasted %d of %d %s (%.1fM of %.1fM)\n",
	    space-used,
	    space,
	    (space == 1 ? "block" : "blocks"),
	    (space-used)*DEFAULT_BLOCK_SIZE / 1048576.0,
	    (space*DEFAULT_BLOCK_SIZE) / 1048576.0);

    free(results);
    free(weights);
    free(nonopts);

    exit(0);
}
